### ISSUE TITLE
Error: Could not remove initial owner from service principal with object ID

### ISSUE THREAD

**Issue by jtravell** [hashicorp/terraform-provider-azuread#529]
An error is being thrown when we use the very simple code below gotten mainly from the official terraform/azuread documentation.  We are tagging this as a bug, given that this example is so simple and uses code from official documentation.  
  
>**What specific syntax must be changed in the specific terraform code given at bottom below in order to resolve the `Error: Could not remove initial owner from service principal with object ID` error message which is currently being thrown?**  
  
**COMMAND THAT TRIGGERS ERROR:**  
  
The following command triggers the error message below it when the following command invokes the terraform at bottom below.  
    
    terraform apply -var="instanceName=tfBackend-myOrg" -var="resourceGroupName=admin" -var="resourceGroupRegion=westus" -var-file="C:\path\to\keys.tfvars"
  
The contents of the `keys.tfvars` referred to in the above command are:  
  
    subscriptionId="actual-subscription-id-redacted"
    tenantId="actual-tenant-id-redacted"
    clientId="actual-client-id-redacted"
    clientSecret="actual-secret-redacted"
  
**CONFIGURATION:**  
  
The provider version given in the `terraform.lock.hcl` file is as follows:  
  
    provider "registry.terraform.io/hashicorp/azuread" {
      version = "2.0.1"
      hashes = [ hashes redacted for brevity/security ]
    }
  
The version of terraform is `v1.0.5`.  
  
The roles assigned to the `clientId` used to run the terraform command above include:  
* Subscription: Owner  
* Active Directory: Global Administrator  
  
**ERROR LOG:**  
  
The complete error log is as follows:  
  
    module.ad-admin.azuread_service_principal.appRegistrationSP: Still creating... [2m0s elapsed]
    2021/08/27 12:18:42 [DEBUG] module.ad-admin.azuread_service_principal.appRegistrationSP: apply errored, but we're indicating that via the Error pointer rather than returning it: Could not remove initial owner from service principal with object ID: "guid-alpha-numeric-string": ServicePrincipalsClient.BaseClient.Get(): Get "https://graph.microsoft.com/beta/tenant-id-alpha-numeric-string/servicePrincipals/guid-alpha-numeric-string/owners/guid-alpha-numeric-string/$ref?%24select=id%2Curl": GET https://graph.microsoft.com/beta/tenant-id-alpha-numeric-string/servicePrincipals/guid-alpha-numeric-string/owners/guid-alpha-numeric-string/$ref?%24select=id%2Curl giving up after 9 attempt(s)
    2021/08/27 12:18:42 [TRACE] EvalMaybeTainted: module.ad-admin.azuread_service_principal.appRegistrationSP encountered an error during creation, so it is now marked as tainted
    2021/08/27 12:18:42 [TRACE] EvalWriteState: recording 1 dependencies for module.ad-admin.azuread_service_principal.appRegistrationSP
    2021/08/27 12:18:42 [TRACE] EvalWriteState: writing current state object for module.ad-admin.azuread_service_principal.appRegistrationSP
    2021/08/27 12:18:42 [TRACE] EvalApplyProvisioners: azuread_service_principal.appRegistrationSP is tainted, so skipping provisioning
    2021/08/27 12:18:42 [TRACE] EvalMaybeTainted: module.ad-admin.azuread_service_principal.appRegistrationSP was already tainted, so nothing to do
    2021/08/27 12:18:42 [TRACE] EvalWriteState: recording 1 dependencies for module.ad-admin.azuread_service_principal.appRegistrationSP
    2021/08/27 12:18:42 [TRACE] EvalWriteState: writing current state object for module.ad-admin.azuread_service_principal.appRegistrationSP
    2021/08/27 12:18:42 [TRACE] vertex "module.ad-admin.azuread_service_principal.appRegistrationSP": visit complete
    2021/08/27 12:18:42 [TRACE] dag/walk: upstream of "module.ad-admin.azuread_service_principal_password.appRegistrationSP_pwd" errored, so skipping
    2021/08/27 12:18:42 [TRACE] dag/walk: upstream of "module.ad-admin.provider[\"registry.terraform.io/hashicorp/azuread\"] (close)" errored, so skipping
    2021/08/27 12:18:42 [TRACE] ExecuteWriteOutput: Saving Create change for module.ad-admin.output.application_id in changeset
    2021/08/27 12:18:42 [TRACE] EvalWriteOutput: Saving value for module.ad-admin.output.application_id in state
    2021/08/27 12:18:42 [TRACE] vertex "module.ad-admin.output.application_id": visit complete
    2021/08/27 12:18:42 [TRACE] vertex "module.ad-admin.output.application_id (expand)": dynamic subgraph completed successfully
    2021/08/27 12:18:42 [TRACE] vertex "module.ad-admin.output.application_id (expand)": visit complete
    2021/08/27 12:18:42 [TRACE] dag/walk: visiting "output.application_id"
    2021/08/27 12:18:42 [TRACE] vertex "output.application_id": starting visit (*terraform.NodeApplyableOutput)
    2021/08/27 12:18:42 [TRACE] dag/walk: upstream of "module.ad-admin (close)" errored, so skipping
    2021/08/27 12:18:42 [TRACE] ExecuteWriteOutput: Saving NoOp change for output.application_id in changeset
    2021/08/27 12:18:42 [TRACE] EvalWriteOutput: Saving value for output.application_id in state
    2021/08/27 12:18:42 [TRACE] vertex "output.application_id": visit complete
    2021/08/27 12:18:42 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
    2021/08/27 12:18:42 [TRACE] dag/walk: upstream of "root" errored, so skipping
    2021/08/27 12:18:42 [TRACE] statemgr.Filesystem: have already backed up original terraform.tfstate to terraform.tfstate.backup on a previous write
    2021/08/27 12:18:42 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 14
    2021/08/27 12:18:42 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
    2021/08/27 12:18:42 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
    2021/08/27 12:18:42 [TRACE] statemgr.Filesystem: unlocked by closing terraform.tfstate
    2021-08-27T12:18:42.831-0700 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
    
    Error: Could not remove initial owner from service principal with object ID: "guid-alpha-numeric-string"
    
      on ..\..\..\..\modules\ad-admin\active-directory.tf line 65, in resource "azuread_service_principal" "appRegistrationSP":
      65: resource "azuread_service_principal" "appRegistrationSP" 2021-08-27T12:18:42.856-0700 [DEBUG] plugin: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/azuread/2.0.1/windows_amd64/terraform-provider-azuread_v2.0.1_x5.exe pid=3556
    2021-08-27T12:18:42.856-0700 [DEBUG] plugin: plugin exited
    {
    
    ServicePrincipalsClient.BaseClient.Get(): Get
    "https://graph.microsoft.com/beta/tenant-id-alpha-numeric-string/servicePrincipals/guid-alpha-numeric-string/owners/guid-alpha-numeric-string/$ref?%24select=id%2Curl":
    GET
    https://graph.microsoft.com/beta/tenant-id-alpha-numeric-string/servicePrincipals/guid-alpha-numeric-string/owners/guid-alpha-numeric-string/$ref?%24select=id%2Curl
    giving up after 9 attempt(s)
    
NOTE:  When we subsequently type `az ad sp show --id guid-alpha-numeric-string` in a cloud shell in the Azure Portal, the result is, as expected, a JSON summary of the service principal that was created by the below code while the above command was running.  The service principal that is created by running the code in this ticket has `"objectId":  "guid-alpha-numeric-string"`.  So that the code seems to create the service principal but then is not able to "remove initial owner from service principal", as stated in the error message shown above.    
    
**COMPLETE CODE TO REPRODUCE THE ERROR:**
    
The complete code for reproducing this error is as follows:  
    
    ###############################################################################################
    ### Input Variables 
    ###############################################################################################
   
    variable "subscriptionId" { }
    variable "tenantId" { }
    variable "clientId" { }
    variable "clientSecret" { }
    variable "instanceName" { }
    
    ###############################################################################################
    ### Providers 
    ###############################################################################################
    
    # Configure the Microsoft Azure Active Directory Provider
    provider "azuread" {
      tenant_id       = var.tenantId
      client_id       = var.clientId
      client_secret   = var.clientSecret
    }
    
    ################################################################################################
    ### Identity Resources
    ################################################################################################
    
    # Create an application
    resource "azuread_application" "appRegistration" {
      display_name = var.instanceName
      app_role {
        allowed_member_types  = ["User", "Application"]
        description           = "Admins can manage roles and perform all task actions"
        display_name          = "Admin"
        value                 = "administer"
        id                    = "1b19509b-32b1-4e9f-b71d-4992aa991967"
      }
    }
    
    # Create a service principal
    resource "azuread_service_principal" "appRegistrationSP" {
      application_id = azuread_application.appRegistration.application_id
    }
    
    resource "azuread_service_principal_password" "appRegistrationSP_pwd" {
      service_principal_id = azuread_service_principal.appRegistrationSP.id
    }

**Comment by croeck** [comment]
Having the same issue here. Tested with Terraform `v1.0.2` and `v1.0.5` running on either Mac OS X or in an isolated Docker container on a linux host. My logs don't show any other insightful information than what was already shared.  Ping me if I can help to track this down.

**Comment by manicminer** [comment]
Hi @jtravell

Thanks for reporting this issue. This looks likely to be an eventual consistency issue - in your log the provider has tried 9 times to retrieve the service principal owners after initially creating it, before giving up, i.e. the service principal is created successfully but there is a delay before the API reflects this. I have not been able to reproduce this right now, but I have seen this before at periods of high load and/or heavy delays in AAD.

We are hoping to add additional mitigation for these kinds of service delays and consistency issues. In the meantime I would suggest simply retrying the apply operation, and consider raising an Azure support ticket if the delays are persistent or excessive.

**Comment by manicminer** [comment]
Note that I have also seen this when a tenant has just had a lot of new objects created - the more objects you create in a short period the longer the delays often seem to get.

For context, the provider is configured to retry up to 9 times by default for a given operation, or until the timeout expires, whichever is sooner.

**Comment by jtravell** [comment]
@manicminer Retrying the `apply` several times did not resolve the problem. We have continued to retry now for 20 minutes, which should far exceed any reasonable estimate of the possible latency problems that you suspected in your comment above. 

Note that the error message states `Error: Could not remove initial owner from service principal with object ID:` .  

The error message does not say anything about being unable to find the new object as you suggest.  Instead, the error message says that the azuread provider is not able to **OPERATE ON** the newly created object, which is entirely different.  

Also note that another user was able to recreate this as shown in the comments above.  

The reproducibility of this problem indicates that it is more than simply the latency problem you hypothesize.

**Comment by manicminer** [comment]
@jtravell That error message wraps more than one operation including a Get operation on the SP owners (which is indicated in the first trace you posted); this is the operation that is failing, almost certainly due to eventual consistency as the trace indicates the SP ID and the owner ID (which is the calling principal) are present in the URL.

**Comment by manicminer** [comment]
@jtravell If you are seeing the exact same error as in your initial report, one thing you could do to sidestep this part of the Create operation is to specify your calling principal as the owner of the SP you are trying to create, e.g.

```hcl
data "azuread_client_config" "test" {}

resource "azuread_service_principal" "test" {
  application_id = azuread_application.test.application_id
  owners = [data.azuread_client_config.test.object_id]
}
```

This will cause the provider to _not_ attempt to remove the calling principal after initially creating the SP. It's not a solution, but a potential interim workaround.

**Comment by manicminer** [comment]
@jtravell, @croeck I've pushed a test build to manicminer/terraform-provider-azuread on the TF registry - this has extra configurable control over API retries. It would be really useful if you could spare a few minutes to test and see if you are able to get past this error by dialing up the `retry_limit` (the max allowed value is 100).

e.g.
```hcl
terraform {
  required_providers {
    azuread = {
      source = "manicminer/azuread"
      version = "11.2.0"
    }
  }
}

provider "azuread" {
  retry_limit = 9 # this is the default
}

resource "azuread_application" "test" {
  display_name = "test"
}

resource "azuread_service_principal" "test" {
  application_id = azuread_application.test.application_id
}

resource "azuread_service_principal_password" "test" {
  service_principal_id = azuread_service_principal.test.object_id
}
```

Note: this is a test build which you shouldn't use in production as has not been peer reviewed and will be removed later

**Comment by jtravell** [comment]
@manicminer  , @croeck   

The following workaround causes the error to go away:  

    data "azuread_client_config" "test" {}

    # Create an application
    resource "azuread_application" "appRegistration" {
      display_name = var.instanceName
      app_role {
        allowed_member_types  = ["User", "Application"]
        description           = "Admins can manage roles and perform all task actions"
        display_name          = "Admin"
        value                 = "administer"
        id                    = "1b19509b-32b1-4e9f-b71d-4992aa991967"
      }
    }

I will be happy to try the test build you mentioned later when I am back at my computer again.  

Thank you for the test build that we will test and also for the workaround that I am showing here as working.  

Note however that retrying the `apply` an hour and ten minutes later just now continued to reproduce the same error in the OP.  We left the `apply` in place for 70 minutes, meaning that the infrastructure was still in place for the entire 70 minutes, and then ran `apply` again and received the same exact error.  I am skeptical that dialing up the `retry_limit` would mitigate a problem that is not resolved by waiting 70 minutes as we did just now.  

# Create a service principal
resource "azuread_service_principal" "appRegistrationSP" {
  application_id = azuread_application.appRegistration.application_id
  owners = [data.azuread_client_config.test.object_id]
}

**Comment by croeck** [comment]
@manicminer I gave your `manicminer/azuread` release a try and set the retry number to 99. The step was now running for 10 minutes until we hit `context deadline exceeded`.

I tested the creation of the service principal multiple times at different hours of the day, so I would say a "busy API" is rather unlikely.

**Comment by manicminer** [comment]
Thanks both for your feedback.

With the additional retry time I agree that service delays are unlikely to be a root cause. I wonder if the initial ownership is not being honored and the subsequent assumption by the provider is then invalid. Could either send me a full debug log with all requests? A gist would be great, if you want to keep it private you can find my email address in my got commits. Also if you destroy all resources afterwards then you'd need only censor the tenant ID from the traces.

**Comment by graememeyer** [comment]
I'm getting this issue as well, following the [Azure Active Directory Provider ~docs](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs). 

My exact code:

``` terraform
# Configure the Azure Active Directory Provider
provider "azuread" {
  tenant_id = "TENANTIDREDACTED"
}

# Retrieve domain information
data "azuread_domains" "example" {
  only_initial = true
}

# Create an application
resource "azuread_application" "example" {
  display_name = "ExampleApp"
}

# Create a service principal
resource "azuread_service_principal" "example" {
  application_id = azuread_application.example.application_id
}

# Create a user
resource "azuread_user" "example" {
  user_principal_name = "ExampleUser@${data.azuread_domains.example.domains.0.domain_name}"
  display_name        = "Example User"
  password            = "PASSWORDREDACTED"
}
```

Terraform version:
```
erraform --version
Terraform v1.0.4
on windows_amd64
+ provider registry.terraform.io/hashicorp/azuread v2.0.1

Your version of Terraform is out of date! The latest version
is 1.0.5. You can update by downloading from https://www.terraform.io/downloads.html
```

I've e-mailed my trace log to your [at] * [dot] io address - couldn't see a hashicorp address anywhere in your git commits.

**Comment by graememeyer** [comment]
I did a `terraform destroy` subsequent to the traced events, and it ran successfully, but failed to delete the newly created example user:

![image](https://user-images.githubusercontent.com/3931530/131258272-e624d3d5-1f02-4399-8197-66cd95030e64.png)

**Comment by graememeyer** [comment]
Sorry one more bit of info. It looks like the userID given in the error is _not_ the one the object is given in azure at the end of the run. 

E.g. I get error:

``` powershell
Error: Could not remove initial owner from service principal with object ID: "object-id-1"
│
│   with azuread_service_principal.example,
│   on main.tf line 17, in resource "azuread_service_principal" "example":
│   17: resource "azuread_service_principal" "example" {
...
```

But when I go to the user properties in Azure to check, I get "object-id-somethingelse"

![image](https://user-images.githubusercontent.com/3931530/131258749-dd5128be-3c94-441e-afc4-6d0e51c47e8f.png)

Both the ID given in the error and the ID listed in azure are mentioned in the trace logs, but I don't understand enough about this yet to say what's happening. I'll keep reading.

**Comment by graememeyer** [comment]
Tried a couple more things, and I think I've determined that the ServicePrincipal/Enterprise Application either isn't given an owner, or when the owner is removed, terraform-provider-azuread isn't able to tell. After the remove initial owner operation fails, the Service Principal doesn't have an owner:

``` 
azuread_service_principal.example: Still creating... [2m0s elapsed]
╷
│ Error: Could not remove initial owner from service principal with object ID: "27157929-7c8b-482e-a254-f4eed83e701e"
│
│   with azuread_service_principal.example,
│   on main.tf line 17, in resource "azuread_service_principal" "example":
│   17: resource "azuread_service_principal" "example" {
│
│ ServicePrincipalsClient.BaseClient.Get(): Get "https://graph.microsoft.com/beta/TENANT_ID_REDACTED/servicePrincipals/27157929-7c8b-482e-a254-f4eed83e701e/owners/bd965c60-d746-45e9-a7f1-c9ca2662a0af/$ref?%24select=id%2Curl":    
│ GET https://graph.microsoft.com/beta/TENANT_ID_REDACTED/servicePrincipals/27157929-7c8b-482e-a254-f4eed83e701e/owners/bd965c60-d746-45e9-a7f1-c9ca2662a0af/$ref?%24select=id%2Curl giving up after 9 attempt(s)
╵
```

![image](https://user-images.githubusercontent.com/3931530/131319338-c8ee3a18-ca31-4210-b385-98244f5f37b6.png)

If I then add an owner manually:

![image](https://user-images.githubusercontent.com/3931530/131319450-960e2753-6450-48e1-908d-84d225904140.png)

I was able to check the Enterprise Application early in the creation process and it didn't appear to have an owner, so I think it's not being given an owner in the first place.

**Comment by graememeyer** [comment]
Tested a few more ways with the AzureAD PowerShell library and the GUI, and it seems that neither Applications/App Registrations, not Enterprise Applications/ServicePrincipals are given an owner by default when you create them. The bug would seem to me to be that when terraform apply creates these resources, there's no need to remove a non-existant owner, unless there should be one created due to a quirk of how Terraform operates (I'm new to TF and for sure have little idea what I'm doing).

**Comment by graememeyer** [comment]
Interestingly, I've just tried specifying an owner for the Service Principal, but TF seems to have failed to add it.

``` Terraform
...
# Retrieve domain information
data "azuread_domains" "example" {
  only_initial = true
}

# Bug fix
data "azuread_client_config" "current" {}

# Create an application
resource "azuread_application" "example" {
  display_name = "ExampleApp"
  owners       = [data.azuread_client_config.current.object_id]
}

# Create a service principal
resource "azuread_service_principal" "example" {
  application_id = azuread_application.example.application_id
  owners         = [data.azuread_client_config.current.object_id]
}
...
```

``` Terraform
Terraform will perform the following actions:

  # azuread_application.example will be created
  + resource "azuread_application" "example" {
      + app_role_ids                = (known after apply)
      + application_id              = (known after apply)
      + disabled_by_microsoft       = (known after apply)
      + display_name                = "ExampleApp"
      + id                          = (known after apply)
      + logo_url                    = (known after apply)
      + oauth2_permission_scope_ids = (known after apply)
      + object_id                   = (known after apply)
      + owners                      = [
          + "REDACTED_OBJECT_ID_1",
        ]
      + prevent_duplicate_names     = false
      + publisher_domain            = (known after apply)
      + sign_in_audience            = "AzureADMyOrg"
    }

  # azuread_service_principal.example will be created
  + resource "azuread_service_principal" "example" {
      + account_enabled             = true
      + app_role_ids                = (known after apply)
      + app_roles                   = (known after apply)
      + application_id              = (known after apply)
      + application_tenant_id       = (known after apply)
      + display_name                = (known after apply)
      + homepage_url                = (known after apply)
      + id                          = (known after apply)
      + logout_url                  = (known after apply)
      + oauth2_permission_scope_ids = (known after apply)
      + oauth2_permission_scopes    = (known after apply)
      + object_id                   = (known after apply)
      + owners                      = [
          + "REDACTED_OBJECT_ID_1",
        ]
      + redirect_uris               = (known after apply)
      + saml_metadata_url           = (known after apply)
      + service_principal_names     = (known after apply)
      + sign_in_audience            = (known after apply)
      + type                        = (known after apply)
    }
...
```

``` terraform
...
azuread_application.example: Creating...
...
azuread_application.example: Creation complete after 1s [id=ac10ac2f-5125-4ee0-8486-432ccc4dbbd2]        
azuread_service_principal.example: Creating...
azuread_service_principal.example: Creation complete after 0s [id=69aae805-7294-4a46-88ba-f31fd7953c45]  
```

The application is given an owner, but the service principal is not:

``` PowerShell
PS > Get-AzureADApplicationOwner -ObjectID ac10ac2f-5125-4ee0-8486-432ccc4dbbd2

ObjectId                             DisplayName  UserPrincipalName                                               UserT
                                                                                                                  ype
--------                             -----------  -----------------                                               -----
REDACTED_OBJECT_ID_1 Graeme Meyer REDACTED.com#EXT#@REDACTED.onmicrosoft.com Me...


PS > Get-AzureADServicePrincipalOwner -ObjectId 69aae805-7294-4a46-88ba-f31fd7953c45
PS > 
```

**Comment by jtravell** [comment]
@manicminer   The following workaround, which you gave above, is failing when we run the same terraform module a second time from a different `.tfstate` using a different value for the `instanceName` input variable, but using the same azure credentials to run terraform:  

    data "azuread_client_config" "test" {}
    
    resource "azuread_service_principal" "appRegistrationSP" {
      application_id = azuread_application.appRegistration.application_id
      owners = [data.azuread_client_config.test.object_id]
    }

In other words, we use one azure set of credentials to run the same terraform module separately from multiple different directories that each have their own separate `.tfstate` files.  And each time that we separately the same terraform module from different directories managed under different `.tfstate`, we use different values for the `instanceName` input variable, meaning different name for the service principal.  ...  
  
**Before adding your workaround, we were able to successfully create N-instances of the resulting service principal, with one instance created to correspond with each of the `.tfstate` files.**  ...  
  
But after adding your workaround, the 2nd through Nth instances are now NOT visible in the Azure Portal Web UI even though the 2nd through Nth runs of the `terraform apply` that received different `instanceName` values as input and that were each created under separate `.tfstate` each report `Apply Complete`.  

**So the Azure Portal UI does NOT show the 2nd through Nth instances, apparently because the `owners = [data.azuread_client_config.test.object_id]` line that you added as a workaround fails to allow for individually-named instances of service principals to be created with multiple separate instances of the same module using different `instanceName` values and different `.tfstate` files.**  ...  In addition, the 2nd through Nth instances of other types of Azure resources outside the scope of this OP are also not visible anymore as a result of your workaround failing to make the 2nd through Nth instances of the service principal visible in the Azure Portal UI.  For example, the modules create a key vault with a secret containing the password for each of the service principals, but you workaround make these 2nd through Nth instances not visible either in the Azure UI Portal.  

The workaround that you suggested does not meet our use case.

We have to be able to create an arbitrary number of service principals which must be associated with an arbitrary number of other companion resources such as key vaults bundled into a terraform module.  ...  Our use case was satisfied in old versions of the provider.  But this bug we are reporting still blocks our use case from being met.  

Your workaround only works if we run the terraform module one time to create only one instance.  

>**Can you please suggest a better workaround that supports being run multiple times with different values for `instanceName` each time?**

**Comment by jtravell** [comment]
@croeck  @graememeyer  Are you willing to please run the same module from a few separate directories using a different value for `instanceName` from each directory from which you call the otherwise same module?  

This test would result in a separate `.tfstate` file being created in each calling directory.  

And could you please then look in the Azure Portal Web UI to see if the 2nd through Nth instances are visible after you create them?  

This test succeeded with older versions of the `azuread` provider, but now is failing due to the workaround suggested in this OP.  

My comment immediately above this gives a detailed explanation of what is failing.  

The only thing I would add is that if we delete each instance before creating the `n+1th` instance, then we see a new value for `instanceName` each time.  However, when we keep each prior instance alive when we run `terraform apply` to create each ne `n+1th` instance, we experience the failed result of the resulting service principal and key vaults, etc NOT being visible in the UI, because it seems that the workaround given is ignoring the separate `.tfstate` for each instance and instead seems to be treating it all as one instance, which is NOT what is required.

**Comment by graememeyer** [comment]
@jtravell I don't use an instance name in my own code, and I think what you're describing is the same issue I mentioned at the end of [my previous comment](https://github.com/hashicorp/terraform-provider-azuread/issues/529#issuecomment-908276246) - the _first_ time you run `terraform apply`, the service principal is not created. But if you run it again, it is created.

**Comment by manicminer** [comment]
@jtravell, @croeck, @graememeyer I've published a test build to manicminer/azuread v11.3.0 which fixes the marshaling bug with service principal initial owners. Could you give this a try and let me know if it works for you (with or without setting `owners` in your config)? Thanks!

**Comment by graememeyer** [comment]
How do you actually load a provider manually for testing? For the life of me I can't get it to work.

I've created a terraform rc file here:

`"C:\Users\username\AppData\Roaming\terraform.rc"`

Which contains:

``` Terraform
provider_installation {
  filesystem_mirror {
    path    = "C:\\Users\\username\\terraform\\azure-ad-terraform\\.terraform\\providers"
    include = ["example.com/*/*"]
  }
    direct {
    exclude = ["example.com/*/*"]
  }
}
```

And I've updated main.tf to include:

``` Terraform
# Configure Terraform
terraform {
  required_providers {
    azuread = {
      source  = "example.com/examplecorp/azuread"
      version = "~> 11.3.0"
    }
  }
}
```

I've placed the provider at path:

"C:\Users\username\terraform\azure-ad-terraform\.terraform\providers\example.com\examplecorp\azuread\11.3.0\windows_amd64\terraform-provider-azuread_v11.3.0.exe"

But I'm getting error:

``` Terraform
terraform init

Initializing the backend...

Initializing provider plugins...
- Finding example.com/examplecorp/azuread versions matching "~> 11.3.0"...
- Installing example.com/examplecorp/azuread v11.3.0...
╷
│ Error: Failed to install provider
│
│ Error while installing example.com/examplecorp/azuread v11.3.0: cannot install existing provider
│ directory .terraform/providers/example.com/examplecorp/azuread/11.3.0/windows_amd64 to itself
```

**Comment by manicminer** [comment]
@graememeyer As it's published to the TF REgistry, there's no need to edit terraformrc and you can just swap out

```hcl
terraform {
  required_providers {
    azuread = {
      source  = "hashicorp/azuread"
      version = "~> 2.0.0"
    }
  }
}
```

with

```hcl
terraform {
  required_providers {
    azuread = {
      source  = "manicminer/azuread"
      version = "11.3.0"
    }
  }
}
```

in your module config

**Comment by graememeyer** [comment]
Ahhh thankyou! I would never have figured that out on my own 🤦‍♀️

Without specifying an owner, it runs quickly, "succeeds", and does not apply an owner to the objects:

``` Terraform
# Configure Terraform
terraform {
  required_providers {
    azuread = {
      source  = "manicminer/azuread"
      version = "11.3.0"
    }
  }
}

# Configure the Azure Active Directory Provider
provider "azuread" {
  tenant_id = "REDACTED_TENANT_ID"
}

# Retrieve domain information
data "azuread_domains" "example" {
  only_initial = true
}

# Bug fix
data "azuread_client_config" "current" {}

# Create an application
resource "azuread_application" "example" {
  display_name = "ExampleApp"
  # owners       = [data.azuread_client_config.current.object_id]
}

# Create a service principal
resource "azuread_service_principal" "example" {
  application_id = azuread_application.example.application_id
  # owners         = [data.azuread_client_config.current.object_id]
}
```

``` Terraform
Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

azuread_application.example: Creating...
azuread_application.example: Creation complete after 1s [id=bab1c238-66b6-41d1-bcd7-bac50da46e1d]
azuread_service_principal.example: Creating...
azuread_service_principal.example: Creation complete after 1s [id=04127a08-fea1-49d9-96f3-1c9707167eac] 
```

``` PowerShell
PS > Get-AzureADApplication -ObjectId bab1c238-66b6-41d1-bcd7-bac50da46e1d        

ObjectId                             AppId                                DisplayName
--------                             -----                                -----------
bab1c238-66b6-41d1-bcd7-bac50da46e1d 943af41d-b142-4041-ae6b-4e9cb3819715 ExampleApp


PS > Get-AzureADApplicationOwner -ObjectId bab1c238-66b6-41d1-bcd7-bac50da46e1d
PS > Get-AzureADServicePrincipal -ObjectId 04127a08-fea1-49d9-96f3-1c9707167eac

ObjectId                             AppId                                DisplayName
--------                             -----                                -----------
04127a08-fea1-49d9-96f3-1c9707167eac 943af41d-b142-4041-ae6b-4e9cb3819715 ExampleApp


PS > Get-AzureADServicePrincipalOwner -ObjectId 04127a08-fea1-49d9-96f3-1c9707167eac
```

`terraform destroy` also runs quickly, succeeds, and appears to have successfully deleted everything.

I confirm also that when the "owners" lines are un-commented, TF/azuread successfully creates the objects with the owner on the first try, and successfully deletes them all again.

Great work! Thank you ♥

**Comment by jtravell** [comment]
@manicminer   

To reproduce the complete problem, simply do the following:  

First, place the same exact code from the OP above into some directory as follows:  

    ############################################################################################
    ### Input Variables 
    ############################################################################################

    variable "subscriptionId" { }
    variable "tenantId" { }
    variable "clientId" { }
    variable "clientSecret" { }
    variable "instanceName" { }

    ############################################################################################
    ### Providers 
    ############################################################################################

    # Configure the Microsoft Azure Active Directory Provider
    provider "azuread" {
      tenant_id       = var.tenantId
      client_id       = var.clientId
      client_secret   = var.clientSecret
    }

    #############################################################################################
    ### Identity Resources
    #############################################################################################

    # Create an application
    resource "azuread_application" "appRegistration" {
      display_name = var.instanceName
      app_role {
        allowed_member_types  = ["User", "Application"]
        description           = "Admins can manage roles and perform all task actions"
        display_name          = "Admin"
        value                 = "administer"
        id                    = "1b19509b-32b1-4e9f-b71d-4992aa991967"
      }
    }

    # Create a service principal
    resource "azuread_service_principal" "appRegistrationSP" {
      application_id = azuread_application.appRegistration.application_id
    }

    resource "azuread_service_principal_password" "appRegistrationSP_pwd" {
      service_principal_id = azuread_service_principal.appRegistrationSP.id
    }

Second, create a module to call that above code as follows, and copy this call into two different directories so that each directory will get its own `.tfstate` created:  
  
    module "name-of-this-module" {
      source = "..\\..\\..\\modules\\name-of-this-module"

      subscriptionId       = var.subscriptionId
      tenantId             = var.tenantId
      clientId             = var.clientId
      clientSecret         = var.clientSecret
      instanceName         = var.instanceName
      resourceGroupName    = var.resourceGroupName
      resourceGroupRegion  = var.resourceGroupRegion
    }

    ##Input variables.  The client referred to is an App Registration.
    variable "subscriptionId" { }
    variable "tenantId" { }
    variable "clientId" { }
    variable "clientSecret" { }
    variable "instanceName" { }
    variable "resourceGroupName" { }
    variable "resourceGroupRegion" { }

    output "application_id" { value = module.name-of-this-module.application_id }
    output "appId" { value = module.name-of-this-module.appId }

Third, run the following same exact commands from the OP above, separately in each of the two calling directories, BUT MAKE SURE TO USE A DIFFERENT VALUE FOR `instanceName` WHEN YOU CALL FROM EACH DIRECTORY SO THAT EACH CALL WILL CREATE A SEPARATE SERVICE PRINCIPAL.  

    terraform apply -var="instanceName=tfBackend-myOrg" -var="resourceGroupName=admin" -var="resourceGroupRegion=westus" -var-file="C:\path\to\keys.tfvars"
  
You can see the contents of the `keys.tfvars` file in the OP at top.  

The failed result is that only one instance will be created even though you called the module from two different `.tfstate` files with two different names for what should be two different service principals.  
  
**Your proposed changes from `"manicminer/azuread" version = "11.3.0"` DO solve this problem.  However, all production versions of the `azuread` terraform provider since version `1.0.0` have had this bug.**  

We discovered this while upgrading from `0.11.0` to `2.0.1`.  

>**Can you suggest a workaround that allows this use case to work with version `2.0.1` of the `azuread` provider?**  

For now, our only option seems to be to either stick with version `0.11.0` of the provider or to use your `"manicminer/azuread" version = "11.3.0"` , which you said will be destroyed soon.

**Comment by manicminer** [comment]
@jtravell It sounds like you have a bug in your configuration. I spent the time to reproduce as per your steps (commenting out the outputs and unused input vars in the module block) but as expected, two app registrations were created each with a corresponding linked service principal.

![Screenshot 2021-09-01 at 07 30 14](https://user-images.githubusercontent.com/251987/131623093-f43ed215-de8a-4dac-8fcc-faa74941f7ea.png)
![Screenshot 2021-09-01 at 07 30 43](https://user-images.githubusercontent.com/251987/131623101-9120bde9-aa72-4502-81ab-0f8f24f0d2d6.png)

The fix for the original issue will be released in v2.1.0 this week.

Upgrading from 0.11.0 to 2.x is not supported. I recommend upgrading to the latest v1.x before jumping straight to v2.

**Comment by jtravell** [comment]
@manicminer   What version of the azuread provider were you using when trying to recreate the problem?  
  
And what day this week will the changes described in this thread be added into v2.1.0?  
  
The bug is present in the current incarnation of `"hashicorp/azuread" version = "2.1.0"`.  But the bug has been remediated in `"manicminer/azuread" version = "11.3.0"` ,  
 
After the changes in `"manicminer/azuread" version = "11.3.0"` have been released into `"hashicorp/azuread" version = "2.1.0"` , the problem will be resolved and it will not matter whether we agree about the behavior of the iteration of `"hashicorp/azuread" version = "2.1.0"` which triggered this ticket.  

Recall that your initial response way up top was that you did not think the bug in this OP was real and that another user reproduced the issue you were not reproducing.

**Comment by manicminer** [comment]
Hi @jtravell, I'm sorry if you misunderstood my earlier comments, I was just triaging the original bug report and certainly did not dismiss it.

I used 2.0.1 when following your directions for reproducing your secondary issue. Aside from the marshaling bug which is the cause of that original issue, there are no other changes which could account for missing objects. That said, if you believe your issue is resolved then I'm happy the fix is working as expected for you. If not, I'd ask you to open another issue with a reproducible config and I'll be happy to take another look.
